WolvenKit.App/ViewModels/Documents/RDTGraphViewModel.cs

```diff
@@ -69,21 +113,581 @@ public void RenderNodes(CArray<CHandle<graphGraphNodeDefinition>> nodes)
+
+        [Reactive] public string PreviewText { get; set; } = "Drop on connector";
+
+        //protected virtual void OnPreviewTargetChanged()
+        //{
+        //    bool canConnect = PreviewTarget != null && Graph.Schema.CanAddConnection(Source!, PreviewTarget);
+        //    PreviewText = PreviewTarget switch
+        //    {
+        //        SocketViewModel con when con == Source => $"Can't connect to self",
+        //        SocketViewModel con => $"{(canConnect ? "Connect" : "Can't connect")} to {con.Title ?? "pin"}",
+        //        NodeViewModel flow => $"{(canConnect ? "Connect" : "Can't connect")} to {flow.Title ?? "node"}",
+        //        _ => $"Drop on connector"
+        //    };
+        //}
+    }
+
+    public class NodeViewModel : ReactiveObject, INode<SocketViewModel>
+    {
+        [Reactive] public RDTGraphViewModel Graph { get; set; }
+
+        [Reactive] public Point Location { get; set; }
+
+        [Reactive] public bool IsSelected { get; set; }
+
+        public string Header { get; set; }
+        public Dictionary<string, string> Details { get; set; } = new();
+        public string Footer { get; set; }
+
+        public IList<SocketViewModel> Inputs { get; set; } = new ObservableCollection<SocketViewModel>();
+        public IList<SocketViewModel> Outputs { get; set; } = new ObservableCollection<SocketViewModel>();
+
+        public graphGraphNodeDefinition RedNode { get; set; }
+
+        public NodeViewModel()
+        {
+            ((ObservableCollection<SocketViewModel>)Inputs).CollectionChanged += (object sender, NotifyCollectionChangedEventArgs e) =>
+            {
+                if (e.Action == NotifyCollectionChangedAction.Add)
+                {
+                    foreach (SocketViewModel item in e.NewItems)
+                    {
+                        item.Node = this;
+                    }
+                }
+                else if (e.Action == NotifyCollectionChangedAction.Remove)
+                {
+                    foreach (SocketViewModel item in e.OldItems)
+                    {
+                        item.Disconnect();
+                    }
+                }
+            };
+            ((ObservableCollection<SocketViewModel>)Outputs).CollectionChanged += (object sender, NotifyCollectionChangedEventArgs e) =>
+            {
+                if (e.Action == NotifyCollectionChangedAction.Add)
+                {
+                    foreach (SocketViewModel item in e.NewItems)
+                    {
+                        item.Node = this;
+                    }
+                }
+                else if (e.Action == NotifyCollectionChangedAction.Remove)
+                {
+                    foreach (SocketViewModel item in e.OldItems)
+                    {
+                        item.Disconnect();
+                    }
+                }
+            };
+        }
+
+        public virtual Size GetSize()
+        {
+            var size = new Size(200, 0);
```

Comment:
You could bind to the `ActualSize` of the `ItemContainer` if that's what you need here.